### PR TITLE
Fix parser regression in pattern switch parsing

### LINKED ISSUE #25999: When keyword recommender failure with constant named `var`
Some tests are skipped in the branch `features/recursive-patterns` because the compiler treats `var` as a keyword in pattern contexts. It is not clear if that is intentional.

Here is an example from `WhenKeywordRecommenderTests.cs`:

``` c#
        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/25999"), Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
        public async Task TestForSwitchCase_SemanticCheck_AfterClassAndLocalConstantVar_BeforeBreak() =>
    await VerifyKeywordAsync(@"
class var { }
class C
{
    void M() { const object var = null; switch (new object()) { case var $$ break; } }
}");
```

### LINKED ISSUE #26000: Pattern-matching compiler does not permit a constant named `var` as a pattern
See https://github.com/dotnet/roslyn/issues/25999

It isn't clear if this is intentional. If it is, it needs to be documented.

### PR DESCRIPTION
**Pull request by gafter**
`when` is not permitted as a pattern variable name at the top level
Fixes #25999
Fixes #26000
This regression only occurred on the `features/recursive-patterns` branch.

@cston @agocke Please review this small bug fix.
@dotnet/roslyn-compiler FYI

### CODE REVIEW

**Comment by gafter** [comment]
@cston @agocke Can you please review this smallish pattern-matching bug fix?

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs [hunk 1]
```diff
@@ -4475,9 +4460,9 @@ public static void Main(int* a, var* c, Typ* e)
                 // (8,27): error CS0103: The name 'b' does not exist in the current context
                 //             if (a is int* b) {}
                 Diagnostic(ErrorCode.ERR_NameNotInContext, "b").WithArguments("b").WithLocation(8, 27),
-                // (9,22): error CS0244: Neither 'is' nor 'as' is valid on pointer types
+                // (9,22): error CS0119: 'var' is a type, which is not valid in the given context
```
[the remark this answers is not in the document]
Comment (gafter) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs]: Yes, `var` will be a contextual keyword. The decision was made 2018-04-04 11:32am Pacific time. See [https://github.com/dotnet/csharplang/issues/1054#issuecomment-378717038](https://github.com/dotnet/csharplang/issues/1054#issuecomment-378717038)

---
In reply to: [180930247](https://github.com/dotnet/roslyn/pull/26019#discussion_r180930247) [](ancestors = 180930247)

[the remark this answers is not in the document]
Comment (gafter) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests.cs]: However, in the context of this test I believe this is just a multiplication expression `var * d` as pointer types don't work with pattern-matching.

---
In reply to: [180956024](https://github.com/dotnet/roslyn/pull/26019#discussion_r180956024) [](ancestors = 180956024,180930247)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs [hunk 2]
```diff
@@ -371,6 +372,23 @@ private CSharpSyntaxNode ParseExpressionOrPattern(bool whenIsKeyword, Precedence
             }
         }
 
+        /// <summary>
+        /// Is the current token something that could follow a type in a pattern?
+        /// </summary>
+        /// <returns></returns>
```
[the remark this answers is not in the document]
Comment (gafter) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs]: Fixed.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs [hunk 3]
```diff
@@ -6,6 +6,7 @@
 
 namespace Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax
 {
+    using System;
     using Microsoft.CodeAnalysis.Syntax.InternalSyntax;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs]: Please move `using` statements to outer scope.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs [hunk 4]
```diff
@@ -371,6 +372,22 @@ private CSharpSyntaxNode ParseExpressionOrPattern(bool whenIsKeyword, Precedence
             }
         }
 
+        /// <summary>
+        /// Is the current token something that could follow a type in a pattern?
+        /// </summary>
+        private bool IsTokenFollowingTypeInPattern()
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs]: The name seems to imply looking back to verify the previous tokens represent a type. Perhaps name `CanTokenFollowTypeInPattern` instead.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs (hunk 3 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs]: Deleting `using System;` as it isn't needed. The other one needs to be nested to resolve a conflict with names in the current namespace. It is the same in `LanguageParser.cs`.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs (hunk 4 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs]: Agreed.

**Pull request merged**